File: go/activities/gpu/activities.go
package gpu

import (
	"context"
	"fmt"
	"math/rand"
	"sync/atomic"
	"time"

	"go.temporal.io/sdk/activity"
)

// GPUBoundActivities defines GPU-accelerated activities
type GPUBoundActivities struct {
	gpuCount int
	gpuIndex uint32
}

// NewGPUBoundActivities creates a new instance
func NewGPUBoundActivities(gpuCount int) *GPUBoundActivities {
	if gpuCount <= 0 {
		gpuCount = 2 // Default to 2 GPUs
	}
	return &GPUBoundActivities{
		gpuCount: gpuCount,
		gpuIndex: 0,
	}
}

// EmbeddingResult represents the result of embedding generation
type EmbeddingResult struct {
	DocumentID  string
	Embeddings  [][]float32
	Model       string
	Dimensions  int
	ProcessedAt time.Time
}

// ClassificationResult represents document classification result
type ClassificationResult struct {
	DocumentID   string
	Category     string
	SubCategory  string
	Confidence   float32
	ProcessedAt  time.Time
}

// OCRResult represents OCR processing result
type OCRResult struct {
	DocumentID    string
	ExtractedText string
	Confidence    float32
	Language      string
	ProcessedAt   time.Time
}

// ImageAnalysisResult represents image analysis result
type ImageAnalysisResult struct {
	DocumentID   string
	HasImages    bool
	ImageCount   int
	Features     []string
	ProcessedAt  time.Time
}

// LLMInferenceResult represents LLM inference result
type LLMInferenceResult struct {
	DocumentID  string
	Summary     string
	Keywords    []string
	Sentiment   string
	ProcessedAt time.Time
}

// getNextGPU returns the next GPU to use (round-robin)
func (a *GPUBoundActivities) getNextGPU() string {
	index := atomic.AddUint32(&a.gpuIndex, 1) % uint32(a.gpuCount)
	return fmt.Sprintf("GPU-%d", index)
}

// GenerateEmbeddings generates vector embeddings using GPU
func (a *GPUBoundActivities) GenerateEmbeddings(ctx context.Context, documentID string, chunks []string) (*EmbeddingResult, error) {
	gpu := a.getNextGPU()
	activity.GetLogger(ctx).Info("Generating embeddings on GPU",
		"documentID", documentID,
		"chunks", len(chunks),
		"gpu", gpu)

	// Simulate GPU-accelerated embedding generation
	embeddings := make([][]float32, len(chunks))
	dimensions := 768 // Standard BERT embedding size

	for i := range chunks {
		embedding := make([]float32, dimensions)
		// Simulate embeddings (in real implementation, would use GPU-accelerated model)
		for j := range embedding {
			embedding[j] = rand.Float32()
		}
		embeddings[i] = embedding
		
		// Report heartbeat for long operations
		activity.RecordHeartbeat(ctx, fmt.Sprintf("Generated embedding %d/%d", i+1, len(chunks)))
	}

	return &EmbeddingResult{
		DocumentID:  documentID,
		Embeddings:  embeddings,
		Model:       "sentence-transformers/all-MiniLM-L6-v2",
		Dimensions:  dimensions,
		ProcessedAt: time.Now(),
	}, nil
}

// ClassifyDocument classifies document using GPU-accelerated ML model
func (a *GPUBoundActivities) ClassifyDocument(ctx context.Context, documentID string, content string) (*ClassificationResult, error) {
	gpu := a.getNextGPU()
	activity.GetLogger(ctx).Info("Classifying document on GPU",
		"documentID", documentID,
		"gpu", gpu)

	// Simulate GPU-accelerated classification
	categories := []string{"Technical", "Business", "Legal", "Medical", "Educational"}
	subCategories := []string{"Documentation", "Report", "Contract", "Research", "Manual"}

	// Simulate classification (in real implementation, would use GPU ML model)
	category := categories[rand.Intn(len(categories))]
	subCategory := subCategories[rand.Intn(len(subCategories))]
	confidence := 0.7 + rand.Float32()*0.3

	return &ClassificationResult{
		DocumentID:   documentID,
		Category:     category,
		SubCategory:  subCategory,
		Confidence:   confidence,
		ProcessedAt:  time.Now(),
	}, nil
}

// PerformOCR performs optical character recognition using GPU
func (a *GPUBoundActivities) PerformOCR(ctx context.Context, documentID string, imageData []byte) (*OCRResult, error) {
	gpu := a.getNextGPU()
	activity.GetLogger(ctx).Info("Performing OCR on GPU",
		"documentID", documentID,
		"imageSize", len(imageData),
		"gpu", gpu)

	// Simulate GPU-accelerated OCR
	extractedText := fmt.Sprintf("OCR extracted text from document %s using %s", documentID, gpu)
	confidence := 0.85 + rand.Float32()*0.15

	return &OCRResult{
		DocumentID:    documentID,
		ExtractedText: extractedText,
		Confidence:    confidence,
		Language:      "en",
		ProcessedAt:   time.Now(),
	}, nil
}

// AnalyzeImages analyzes images in document using GPU
func (a *GPUBoundActivities) AnalyzeImages(ctx context.Context, documentID string, documentData []byte) (*ImageAnalysisResult, error) {
	gpu := a.getNextGPU()
	activity.GetLogger(ctx).Info("Analyzing images on GPU",
		"documentID", documentID,
		"gpu", gpu)

	// Simulate GPU-accelerated image analysis
	hasImages := rand.Float32() > 0.3
	imageCount := 0
	features := []string{}

	if hasImages {
		imageCount = rand.Intn(5) + 1
		features = []string{"charts", "diagrams", "photos", "illustrations"}
		// Randomly select some features
		selectedFeatures := []string{}
		for _, feature := range features {
			if rand.Float32() > 0.5 {
				selectedFeatures = append(selectedFeatures, feature)
			}
		}
		features = selectedFeatures
	}

	return &ImageAnalysisResult{
		DocumentID:  documentID,
		HasImages:   hasImages,
		ImageCount:  imageCount,
		Features:    features,
		ProcessedAt: time.Now(),
	}, nil
}

// RunLLMInference runs large language model inference on GPU
func (a *GPUBoundActivities) RunLLMInference(ctx context.Context, documentID string, text string) (*LLMInferenceResult, error) {
	gpu := a.getNextGPU()
	activity.GetLogger(ctx).Info("Running LLM inference on GPU",
		"documentID", documentID,
		"textLength", len(text),
		"gpu", gpu)

	// Simulate GPU-accelerated LLM inference
	summary := fmt.Sprintf("AI-generated summary for document %s: This document contains important information processed by %s.", documentID, gpu)
	
	keywords := []string{"temporal", "workflow", "document", "processing", "pipeline"}
	// Select random keywords
	selectedKeywords := []string{}
	for _, keyword := range keywords {
		if rand.Float32() > 0.4 {
			selectedKeywords = append(selectedKeywords, keyword)
		}
	}

	sentiments := []string{"positive", "neutral", "negative"}
	sentiment := sentiments[rand.Intn(len(sentiments))]

	return &LLMInferenceResult{
		DocumentID:  documentID,
		Summary:     summary,
		Keywords:    selectedKeywords,
		Sentiment:   sentiment,
		ProcessedAt: time.Now(),
	}, nil
}

File: python/activities/gpu/activities.py
import os
import time
import random
import asyncio
from dataclasses import dataclass
from datetime import datetime
from typing import List, Optional
from concurrent.futures import ThreadPoolExecutor

from temporalio import activity
import numpy as np


@dataclass
class EmbeddingResult:
    document_id: str
    embeddings: List[List[float]]
    model: str
    dimensions: int
    processed_at: datetime


@dataclass
class ClassificationResult:
    document_id: str
    category: str
    sub_category: str
    confidence: float
    processed_at: datetime


@dataclass
class OCRResult:
    document_id: str
    extracted_text: str
    confidence: float
    language: str
    processed_at: datetime


@dataclass
class ImageAnalysisResult:
    document_id: str
    has_images: bool
    image_count: int
    features: List[str]
    processed_at: datetime


@dataclass
class LLMInferenceResult:
    document_id: str
    summary: str
    keywords: List[str]
    sentiment: str
    processed_at: datetime


class GPUBoundActivities:
    def __init__(self, gpu_count: int = None):
        self.gpu_count = gpu_count or int(os.environ.get("GPU_COUNT", "2"))
        self.gpu_index = 0
        self.gpu_lock = asyncio.Lock()
        self.executor = ThreadPoolExecutor(max_workers=self.gpu_count)
        activity.logger.info(f"GPU activities initialized with {self.gpu_count} GPUs")

    async def _get_next_gpu(self) -> str:
        """Returns the next GPU to use (round-robin)"""
        async with self.gpu_lock:
            gpu = f"GPU-{self.gpu_index % self.gpu_count}"
            self.gpu_index += 1
            return gpu

    @activity.defn
    async def generate_embeddings(
        self, document_id: str, chunks: List[str]
    ) -> EmbeddingResult:
        """Generates vector embeddings using GPU"""
        gpu = await self._get_next_gpu()
        activity.logger.info(
            f"Generating embeddings on {gpu} for document {document_id}, "
            f"chunks: {len(chunks)}"
        )
        
        # Simulate GPU-accelerated embedding generation
        embeddings = []
        dimensions = 768  # Standard BERT embedding size
        
        for i, chunk in enumerate(chunks):
            # Simulate embeddings (in real implementation, would use GPU-accelerated model)
            # Using numpy for slightly more realistic simulation
            embedding = np.random.randn(dimensions).astype(np.float32)
            # Normalize the embedding
            embedding = embedding / np.linalg.norm(embedding)
            embeddings.append(embedding.tolist())
            
            # Report heartbeat for long operations
            activity.heartbeat(f"Generated embedding {i+1}/{len(chunks)}")
            
            # Simulate GPU processing time
            await asyncio.sleep(0.1)
        
        return EmbeddingResult(
            document_id=document_id,
            embeddings=embeddings,
            model="sentence-transformers/all-MiniLM-L6-v2",
            dimensions=dimensions,
            processed_at=datetime.now()
        )

    @activity.defn
    async def classify_document(
        self, document_id: str, content: str
    ) -> ClassificationResult:
        """Classifies document using GPU-accelerated ML model"""
        gpu = await self._get_next_gpu()
        activity.logger.info(f"Classifying document on {gpu}: {document_id}")
        
        # Simulate GPU-accelerated classification
        categories = ["Technical", "Business", "Legal", "Medical", "Educational"]
        sub_categories = ["Documentation", "Report", "Contract", "Research", "Manual"]
        
        # Simulate classification (in real implementation, would use GPU ML model)
        # Simple heuristic based on content
        category_scores = {}
        for category in categories:
            score = random.random() * 0.5
            if category.lower() in content.lower():
                score += 0.5
            category_scores[category] = score
        
        # Select category with highest score
        category = max(category_scores, key=category_scores.get)
        sub_category = random.choice(sub_categories)
        confidence = category_scores[category]
        
        # Simulate GPU processing time
        await asyncio.sleep(0.2)
        
        return ClassificationResult(
            document_id=document_id,
            category=category,
            sub_category=sub_category,
            confidence=confidence,
            processed_at=datetime.now()
        )

    @activity.defn
    async def perform_ocr(
        self, document_id: str, image_data: bytes
    ) -> OCRResult:
        """Performs optical character recognition using GPU"""
        gpu = await self._get_next_gpu()
        activity.logger.info(
            f"Performing OCR on {gpu} for document {document_id}, "
            f"image size: {len(image_data)} bytes"
        )
        
        # Simulate GPU-accelerated OCR
        extracted_text = f"OCR extracted text from document {document_id} using {gpu}"
        confidence = 0.85 + random.random() * 0.15
        
        # Simulate GPU processing time
        await asyncio.sleep(0.3)
        
        return OCRResult(
            document_id=document_id,
            extracted_text=extracted_text,
            confidence=confidence,
            language="en",
            processed_at=datetime.now()
        )

    @activity.defn
    async def analyze_images(
        self, document_id: str, document_data: bytes
    ) -> ImageAnalysisResult:
        """Analyzes images in document using GPU"""
        gpu = await self._get_next_gpu()
        activity.logger.info(f"Analyzing images on {gpu} for document {document_id}")
        
        # Simulate GPU-accelerated image analysis
        has_images = random.random() > 0.3
        image_count = 0
        features = []
        
        if has_images:
            image_count = random.randint(1, 5)
            possible_features = ["charts", "diagrams", "photos", "illustrations", "tables", "graphs"]
            # Randomly select some features
            num_features = random.randint(1, 4)
            features = random.sample(possible_features, num_features)
        
        # Simulate GPU processing time
        await asyncio.sleep(0.2)
        
        return ImageAnalysisResult(
            document_id=document_id,
            has_images=has_images,
            image_count=image_count,
            features=features,
            processed_at=datetime.now()
        )

    @activity.defn
    async def run_llm_inference(
        self, document_id: str, text: str
    ) -> LLMInferenceResult:
        """Runs large language model inference on GPU"""
        gpu = await self._get_next_gpu()
        activity.logger.info(
            f"Running LLM inference on {gpu} for document {document_id}, "
            f"text length: {len(text)}"
        )
        
        # Simulate GPU-accelerated LLM inference
        summary = (
            f"AI-generated summary for document {document_id}: "
            f"This document contains important information processed by {gpu}. "
            f"The content appears to be related to document processing pipelines."
        )
        
        # Extract some keywords (simulated)
        possible_keywords = [
            "temporal", "workflow", "document", "processing", "pipeline",
            "embeddings", "analysis", "data", "automation", "orchestration"
        ]
        num_keywords = random.randint(3, 6)
        keywords = random.sample(possible_keywords, num_keywords)
        
        # Determine sentiment
        sentiments = ["positive", "neutral", "negative"]
        sentiment_weights = [0.4, 0.5, 0.1]  # Bias towards positive/neutral
        sentiment = random.choices(sentiments, weights=sentiment_weights)[0]
        
        # Simulate GPU processing time (longer for LLM)
        await asyncio.sleep(0.5)
        
        return LLMInferenceResult(
            document_id=document_id,
            summary=summary,
            keywords=keywords,
            sentiment=sentiment,
            processed_at=datetime.now()
        )